### PR TITLE
CSHARP-5436: Optimize special case of Any with constant array and field equals parameter in predicate.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/AllMethodWithContainsInPredicateToFilterTranslator.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@
 
 namespace MongoDB.Driver.Linq.Linq3Implementation.Translators.ExpressionToFilterTranslators.MethodTranslators
 {
-    internal static class AllWithContainsInPredicateMethodToFilterTranslator
+    internal static class AllMethodWithContainsInPredicateToFilterTranslator
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/AllMethodWithContainsInPredicateToFilterTranslator.cs:25]: Just standardizing names of these 4 related classes.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/AllOrAnyMethodToFilterTranslator.cs [hunk 2]
```diff
@@ -32,14 +31,19 @@ internal static class AllOrAnyMethodToFilterTranslator
...
             }
 
-            if (AnyWithContainsInPredicateMethodToFilterTranslator.CanTranslate(expression, out arrayFieldExpression, out arrayConstantExpression))
+            if (AnyMethodWithContainsInPredicateToFilterTranslator.CanTranslate(expression, out arrayFieldExpression, out arrayConstantExpression))
             {
-                return AnyWithContainsInPredicateMethodToFilterTranslator.Translate(context, arrayFieldExpression, arrayConstantExpression);
+                return AnyMethodWithContainsInPredicateToFilterTranslator.Translate(context, arrayFieldExpression, arrayConstantExpression);
+            }
+
+            if (AnyMethodWithConstantArrayAndFieldEqualsParameterInPredicateToFilterTranslator.CanTranslate(expression, out arrayConstantExpression, out var fieldExpression))
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/AllOrAnyMethodToFilterTranslator.cs:44]: This is how we handle special cases in general. We have a translator for the special case, and we ask it whether it recognizes this expression as being the special case.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/AnyMethodWithConstantArrayAndFieldEqualsParameterInPredicateToFilterTranslator.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+        public static AstFilter Translate(TranslationContext context, ConstantExpression arrayConstantExpression, Expression fieldExpression)
+        {
+            var fieldTranslation = ExpressionToFilterFieldTranslator.Translate(context, fieldExpression);
+            var fieldSerializer = fieldTranslation.Serializer;
+
+            var arrayValue = (IEnumerable)arrayConstantExpression.Value;
+            var serializedArrayValue = SerializationHelper.SerializeValues(itemSerializer: fieldSerializer, arrayValue);
+
+            return AstFilter.In(fieldTranslation, serializedArrayValue);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/AnyMethodWithConstantArrayAndFieldEqualsParameterInPredicateToFilterTranslator.cs:77]: There is more code involved to recognize the special case than to handle it... :)

**Comment by sanych-sun** [comment]
As it was discussed offline - we will try to rewrite this PR as AstSimplifier.

**Comment by rstam** [comment]
Ticket reassigned to Alex who will create a new PR.

**Pull request closed without merging**